File: day2/java/Day2.java
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.io.File;

// modified version of day2 to satify part 2
public class Day2 {
    public static File file = new File("input.txt");

    public static void main(String[] args) {
        List<Range> ranges = readInput(file);
        long tallyedInvalidNumbers = 0;
        for (Range range : ranges) {
            List<Long> invalidNumbers = invalidRanges(range);
            for (Long invalidNumber : invalidNumbers) {
                System.out.println("Invalid number: " + invalidNumber);
                tallyedInvalidNumbers += invalidNumber;
            }
        }
        System.out.println("Total invalid numbers: " + tallyedInvalidNumbers);
    }

    private static List<Range> readInput(File file) {
        List<Range> ranges = new ArrayList<>();
        try (Scanner scanner = new Scanner(file)) {
            scanner.useDelimiter(",");
            while (scanner.hasNext()) {
                String line = scanner.next().trim();
                String[] parts = line.split("-");
                long start = Long.parseLong(parts[0]);
                long end = Long.parseLong(parts[1]);
                Range range = new Range(start, end);
                ranges.add(range);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return ranges;
    }

    private static List<Long> invalidRanges(Range range) {
        List<Long> invalidNumbers = new ArrayList<>();
        for (long i = range.getStart(); i <= range.getEnd(); i++) {
            if (isInvalid(i)) {
                invalidNumbers.add(i);
            }
        }
        return invalidNumbers;
    }

    private static boolean isInvalid(long num) {
        String numString = Long.toString(num);
        int len = numString.length();
        for (int blocksize = 1; blocksize < len / 2 + 1; blocksize++) {
            if (len % blocksize !=0) {
                continue;
            }
            String block = numString.substring(0, blocksize);
            boolean repeating = true;
            for (int pos = blocksize; pos < len; pos += blocksize) {
                String currBlock = numString.substring(pos, pos + blocksize);
                if (!currBlock.equals(block)) {
                    repeating = false;
                    break;
                }
            }
            if (repeating) { return true;}
        }
        return false;
    }

    static class Range {
        private long start;
        private long end;

        public Range(long start, long end) {
            this.start = start;
            this.end = end;
        }
        public long getStart() {
            return start;
        }
        public long getEnd() {
            return end;
        }
    }
}

File: day2/day2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var FILE string = "input.txt"

type Range struct {
	start int64
	end   int64
}

func main() {
	ranges := readInput()
	var totaledInvalid int64 = 0
	for _, r := range ranges {
		invalidNumbers := invalidRanges(r)
		for _, num := range invalidNumbers {
			totaledInvalid += num
		}
	}
	fmt.Println(totaledInvalid)
}

func readInput() []Range {
	file, err := os.Open(FILE)
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	line := scanner.Text()

	var ranges []Range
	parts := strings.Split(line, ",")
	for _, part := range parts {
		part = strings.TrimSpace(part)
		rangeParts := strings.Split(part, "-")
		start, _ := strconv.ParseInt(rangeParts[0], 10, 64)
		end, _ := strconv.ParseInt(rangeParts[1], 10, 64)
		ranges = append(ranges, Range{start, end})
	}
	return ranges
}

func invalidRanges(r Range) []int64 {
	var invalidNumbers []int64
	for i := r.start; i <= r.end; i++ {
		if isInvalid(i) {
			invalidNumbers = append(invalidNumbers, i)
		}
	}
	return invalidNumbers
}

func isInvalid(num int64) bool {
	str := strconv.FormatInt(num, 10)
	length := len(str)
	for blocksize := 1; blocksize < length/2+1; blocksize++ {
		if length%blocksize != 0 {
			continue
		}
		block := str[0:blocksize]
		repeating := true
		for pos := blocksize; pos < length; pos += blocksize {
			currBlock := str[pos : pos+blocksize]
			if currBlock != block {
				repeating = false
				break
			}
		}
		if repeating {
			return true
		}
	}
	return false
}
